### PR TITLE
ARM64: Remove new thread entry wrapper and cleanup

### PR DESCRIPTION
**Pull request by carlocaione**
Three commits aiming at reducing and simplifying the assembly code and the context switch routines. Most importantly we are removing the new thread entry wrapper code using instead the general code for context switch. To do that we have to slightly modify the initial stack frame.

We are also introducing a couple of new macros and give a better name to the save / restore context macros.

### CODE REVIEW

**Comment by carlocaione** [comment]
gentle ping about this PR

File: arch/arm/core/aarch64/switch.S
Comment (sandeepbrcm) [inline arch/arm/core/aarch64/switch.S:80]: we will also return to 'isr_wrapper'

File: arch/arm/core/aarch64/isr_wrapper.S
Comment (sandeepbrcm) [inline arch/arm/core/aarch64/isr_wrapper.S:105]: Now that the actual switch takes place at  `z_arm64_exit_exc`  should we elaborate it in comment.  May be somewhere in comment we should mention the flow.

File: arch/arm/core/aarch64/isr_wrapper.S
Comment (sandeepbrcm) [inline arch/arm/core/aarch64/isr_wrapper.S:114]: Now we will always return here right. There is no way out from ` z_arm64_context_switch`
And this seems now much simplified path.

**Review by sandeepbrcm — COMMENTED** [review]
looks good

File: arch/arm/core/aarch64/switch.S
Comment (carlocaione) [reply arch/arm/core/aarch64/switch.S:80]: ACK

File: arch/arm/core/aarch64/isr_wrapper.S
Comment (carlocaione) [reply arch/arm/core/aarch64/isr_wrapper.S:114]: We will return here every time we swap in a task that was preemptively swapped out and yeah, the path is way more clear now.

File: arch/arm/core/aarch64/isr_wrapper.S
Comment (carlocaione) [reply arch/arm/core/aarch64/isr_wrapper.S:105]: ok. I'll also noticed a bunch of other comments that need to be updated. I'll add a comments fixing PR on top of the next respin.

**Review by sandeepbrcm — APPROVED** [review]

**Comment by carlocaione** [comment]
@Abhishek-brcm can you approve this PR if it looks good to you?

File: arch/arm/core/aarch64/macro_priv.inc
Comment (Abhishek-brcm) [inline arch/arm/core/aarch64/macro_priv.inc:131]: I don't understand purpose of xreg2, is that intentional?

Comment (carlocaione) [reply arch/arm/core/aarch64/macro_priv.inc:131]: Not intentional. It is part of a following PR that yet posted based on this PR. I'll fix that.

**Review by tyrantlizard77 — APPROVED** [review]

**Comment by andrewboie** [comment]
@Abhishek-brcm can you take a look once more?

**Comment by Abhishek-brcm** [comment]
> @Abhishek-brcm can you take a look once more?

I will have a look by EoD today.

File: arch/arm/core/aarch64/macro_priv.inc
Comment (carlocaione) [reply arch/arm/core/aarch64/macro_priv.inc:22]: Well, also `z_arm64_{enter,exit}_exc` is not an entirely proper name either since they are being used when entering and exiting the exception but the name is not descriptive of what's going on in the macros

File: arch/arm/core/aarch64/switch.S
Comment (carlocaione) [reply arch/arm/core/aarch64/switch.S:45]: `x30` is already saved as part of the `z_arm64_enter_exc` macro and restored when going through `z_arm64_exit_exc`. 

When `ret` in `z_arm64_context_switch` is taken we are back into `z_arm64_svc` for cooperative swapping or `_isr_wrapper` when preemptively swapping. In both cases before leaving these functions we are restoring `x30` as part of `z_arm64_exit_exc`.

`x30` was saved here as part of the process context because for new threads after `ret` we needed to jump into `z_thread_entry_wrapper` but this is not needed anymore.

File: arch/arm/core/aarch64/macro_priv.inc
Comment (Abhishek-brcm) [reply arch/arm/core/aarch64/macro_priv.inc:22]: I find `z_arm64_{enter,exit}_exc` to be better.. as these (`x0-18, LR, SPSR, ELR`) seem to be the registers which need to be saved on exception/interrupt.. I will leave it to you.

File: arch/arm/core/aarch64/switch.S
Comment (Abhishek-brcm) [reply arch/arm/core/aarch64/switch.S:45]: > `x30` is already saved as part of the `z_arm64_enter_exc` macro and restored when going through `z_arm64_exit_exc`.

I see.

**Review by Abhishek-brcm — APPROVED** [review]

File: arch/arm/core/aarch64/macro_priv.inc
Comment (carlocaione) [reply arch/arm/core/aarch64/macro_priv.inc:22]: No prob. I have removed the renaming commit. Thanks.

**Pull request merged**
